### PR TITLE
catalog: Add sub-operations to transactions

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds sub-operations to catalog transactions. This allows callers to logically group operations within a catalog transaction. This feature will be useful when the catalog is structured to listen to durable changes, because it will allow the catalog to extract durable changes for a single logical operation and apply just those changes to it's in-memory state.

Works towards resolving #24844

### Motivation
This PR adds a known-desirable feature.

### Tips for reviewer

All of the interesting changes are in `TableTransaction` the rest is mostly plumbing.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
This will be used to achieve step (2) of this design: https://materializeinc.slack.com/archives/C058H9MH8P3/p1713454774849519?thread_ts=1713453438.856209&cid=C058H9MH8P3

**Comment by jkosh44** [comment]
This make `TableTransaction`s less efficient, but the DDL benchmark doesn't show any significant changes: https://buildkite.com/materialize/nightly/builds/7542#018f0b9c-ddc5-4aee-9556-dfdaf8fb433f

File: src/ore/src/vec.rs [hunk 1]
```diff
@@ -167,22 +171,27 @@ impl<T> VecExt<T> for Vec<T> {
...
     }
+
+    // implementation is from Vec::is_sorted_by, but with `windows` instead of
+    // the unstable `array_windows`
+    fn is_sorted_by<F>(&self, mut compare: F) -> bool
+    where
+        F: FnMut(&T, &T) -> bool,
+    {
+        self.windows(2).all(|win| compare(&win[0], &win[1]))
+    }
```
Comment (benesch) [inline src/ore/src/vec.rs:182]: It would be good to have a few test cases for this.

**Review by benesch — COMMENTED** [review]
`mz_ore` changes LGTM. Not gonna look at the rest.

File: src/ore/src/vec.rs (hunk 1 above)
Comment (jkosh44) [reply src/ore/src/vec.rs:182]: Done

File: src/catalog/src/durable/transaction.rs [hunk 2]
```diff
@@ -1866,19 +1921,27 @@ where
...
+        );
         Ok(())
     }
 
     /// Iterates over the items viewable in the current transaction in arbitrary
     /// order and applies `f` on all key, value pairs.
     fn for_values<F: FnMut(&K, &V)>(&self, mut f: F) {
         let mut seen = BTreeSet::new();
-        for (k, v) in self.pending.iter() {
+        for (k, _v) in self.pending.iter() {
```
Comment (maddyblue) [inline src/catalog/src/durable/transaction.rs:1938]:
```suggestion
        for k in self.pending.keys() {
```

File: src/catalog/src/durable/transaction.rs [hunk 3]
```diff
@@ -1912,16 +1982,18 @@ where
...
+        for (k, updates) in pending {
+            self.pending.entry(k).or_default().extend(updates);
+        }
     }
 
     /// Inserts a new k,v pair.
     ///
     /// Returns an error if the uniqueness check failed or the key already exists.
-    fn insert(&mut self, k: K, v: V) -> Result<(), DurableCatalogError> {
+    fn insert(&mut self, k: K, v: V, t: u64) -> Result<(), DurableCatalogError> {
```
Comment (maddyblue) [inline src/catalog/src/durable/transaction.rs:1996]: Thoughts on making at least a type alias like `type Timestamp = u64` to make this a tiny bit more explicit? Would help a lot with readability. Diff could get the same treatment.

**Review by maddyblue — APPROVED** [review]
LGTM! All the new logic handling multiple timestamps makes sense and is readable.

File: src/catalog/src/durable/transaction.rs [hunk 4]
```diff
@@ -1785,16 +1851,16 @@ impl TransactionBatch {
 #[derive(Debug, PartialEq, Eq)]
 struct TableTransaction<K, V> {
     initial: BTreeMap<K, V>,
-    // The desired state of keys after commit. `None` means the value will be
-    // deleted.
-    pending: BTreeMap<K, Option<V>>,
+    // The desired updates to keys after commit.
+    // Invariant: Value is sorted by `.1`.
+    pending: BTreeMap<K, Vec<(V, u64, i64)>>,
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/catalog/src/durable/transaction.rs]: Good idea, I converted this to a struct.

File: src/catalog/src/durable/transaction.rs (hunk 2 above)
Comment (jkosh44) [reply src/catalog/src/durable/transaction.rs]: Done

File: src/catalog/src/durable/transaction.rs (hunk 3 above)
Comment (jkosh44) [reply src/catalog/src/durable/transaction.rs]: Good idea, just updated.

**Pull request merged**